File: low-level-design/lld-coding/multi-language/01-basic-classes/bank_account.go
/*
Go Implementation: BankAccount Struct
======================================
Features:
- Struct-based (no classes)
- Constructor function (NewBankAccount)
- Receiver methods
- Private fields (lowercase)
- Public methods (uppercase)
*/

package main

import (
	"fmt"
)

// BankAccount represents a bank account with basic operations
// Note: Fields starting with uppercase are public, lowercase are private
type BankAccount struct {
	AccountNumber string
	balance       float64 // lowercase = private
}

// NewBankAccount creates a new bank account (constructor pattern)
func NewBankAccount(accountNumber string, initialBalance float64) *BankAccount {
	fmt.Printf("Created account: %s\n", accountNumber)
	return &BankAccount{
		AccountNumber: accountNumber,
		balance:       initialBalance,
	}
}

// Deposit adds money to the account
// (ba *BankAccount) is the receiver - like 'self' in Python
func (ba *BankAccount) Deposit(amount float64) bool {
	if amount <= 0 {
		fmt.Println("Error: Deposit amount must be positive")
		return false
	}

	ba.balance += amount
	fmt.Printf("Deposited $%.2f\n", amount)
	return true
}

// Withdraw removes money from the account
func (ba *BankAccount) Withdraw(amount float64) bool {
	if amount <= 0 {
		fmt.Println("Error: Withdrawal amount must be positive")
		return false
	}

	if amount > ba.balance {
		fmt.Printf("Error: Insufficient funds. Balance: $%.2f\n", ba.balance)
		return false
	}

	ba.balance -= amount
	fmt.Printf("Withdrew $%.2f\n", amount)
	return true
}

// GetBalance returns the current balance
func (ba *BankAccount) GetBalance() float64 {
	return ba.balance
}

// String implements the Stringer interface (like __str__ in Python)
func (ba *BankAccount) String() string {
	return fmt.Sprintf("Account %s: $%.2f", ba.AccountNumber, ba.balance)
}

func main() {
	fmt.Println("============================================================")
	fmt.Println("Go: Bank Account Example")
	fmt.Println("============================================================")

	// Create account
	account := NewBankAccount("ACC001", 1000)
	fmt.Printf("Initial balance: $%.2f\n\n", account.GetBalance())

	// Deposit
	account.Deposit(500)
	fmt.Printf("After deposit of $500: $%.2f\n\n", account.GetBalance())

	// Withdraw
	account.Withdraw(200)
	fmt.Printf("After withdrawal of $200: $%.2f\n\n", account.GetBalance())

	// Failed withdrawal
	account.Withdraw(2000)
	fmt.Printf("Final balance: $%.2f\n\n", account.GetBalance())

	// String representation
	fmt.Println("Using String():", account)
}

File: low-level-design/lld-coding/multi-language/01-basic-classes/bank_account.js
/*
 * JavaScript Implementation: BankAccount Class
 * =============================================
 * Features:
 * - ES6 class syntax
 * - Constructor method
 * - Private fields with # (ES2022)
 * - Template literals for strings
 * - Dynamic typing
 */

class BankAccount {
    // Private fields (ES2022 feature)
    #balance;

    /**
     * Constructor - Creates a new bank account
     * @param {string} accountNumber - Unique account identifier
     * @param {number} initialBalance - Starting balance (default: 0)
     */
    constructor(accountNumber, initialBalance = 0) {
        this.accountNumber = accountNumber;
        this.#balance = initialBalance;
        console.log(`Created account: ${accountNumber}`);
    }

    /**
     * Deposit money into account
     * @param {number} amount - Amount to deposit
     * @returns {boolean} - true if successful, false otherwise
     */
    deposit(amount) {
        if (amount <= 0) {
            console.log("Error: Deposit amount must be positive");
            return false;
        }

        this.#balance += amount;
        console.log(`Deposited $${amount.toFixed(2)}`);
        return true;
    }

    /**
     * Withdraw money from account
     * @param {number} amount - Amount to withdraw
     * @returns {boolean} - true if successful, false if insufficient funds
     */
    withdraw(amount) {
        if (amount <= 0) {
            console.log("Error: Withdrawal amount must be positive");
            return false;
        }

        if (amount > this.#balance) {
            console.log(`Error: Insufficient funds. Balance: $${this.#balance.toFixed(2)}`);
            return false;
        }

        this.#balance -= amount;
        console.log(`Withdrew $${amount.toFixed(2)}`);
        return true;
    }

    /**
     * Get current account balance
     * @returns {number} - Current balance
     */
    getBalance() {
        return this.#balance;
    }

    /**
     * String representation (like __str__ in Python)
     * @returns {string}
     */
    toString() {
        return `Account ${this.accountNumber}: $${this.#balance.toFixed(2)}`;
    }
}

// Demo the BankAccount class
function main() {
    console.log("============================================================");
    console.log("JavaScript: Bank Account Example");
    console.log("============================================================");

    // Create account
    const account = new BankAccount("ACC001", 1000);
    console.log(`Initial balance: $${account.getBalance().toFixed(2)}\n`);

    // Deposit
    account.deposit(500);
    console.log(`After deposit of $500: $${account.getBalance().toFixed(2)}\n`);

    // Withdraw
    account.withdraw(200);
    console.log(`After withdrawal of $200: $${account.getBalance().toFixed(2)}\n`);

    // Failed withdrawal
    account.withdraw(2000);
    console.log(`Final balance: $${account.getBalance().toFixed(2)}\n`);

    // String representation
    console.log("Using toString():", account.toString());

    // Try to access private field (will fail)
    console.log("\nTrying to access private field directly:");
    console.log("account.#balance:", typeof account.balance === 'undefined' ? "❌ Cannot access private field" : account.balance);
}

// Run the demo
main();
